### PR TITLE
[Improve][txn] Add admin api updateTransactionCoordinatorNumber

### PR DESCRIPTION
**Pull request by liangyepianzhou**
### Motivation & Modification
 Add admin api updateTransactionCoordinatorNumber to change the partition number of tc

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `no-need-doc` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java [hunk 1]
```diff
@@ -314,4 +315,17 @@ public void getPendingAckInternalStats(@Suspended final AsyncResponse asyncRespo
...
+
+    @POST
+    @Path("/transactionCoordinator/partitions")
+    @ApiResponses(value = {
+            @ApiResponse(code = 503, message = "This Broker is not configured "
+                    + "with transactionCoordinatorEnabled=true."),
+            @ApiResponse(code = 406, message = "The number of partitions should be more than "
+                    + "the current number of transaction coordinator partitions"
+                    + " and less than or equal to maxNumPartitionsPerPartitionedTopic")})
+    public void updateTransactionCoordinatorNumber(int numPartitions) {
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java:327]: It's better to use `scaleTransactionCoordinators(int replicas)`, it's a more user-facing name, and currently, we can only support scale-up.

**Review by codelipenghui — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java [hunk 2]
```diff
@@ -314,4 +315,17 @@ public void getPendingAckInternalStats(@Suspended final AsyncResponse asyncRespo
...
+        checkTransactionCoordinatorEnabled();
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java:328]: should add auth

**Review by congbobo184 — COMMENTED** [review]
could you please add admin client and tools to implement this API?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java (hunk 2 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java:329]: +1, only the super user can make this change.

Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java:329]: I have add the auth just like `updatePartitionedTopic` at last commit.  Are there other problems?

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java [hunk 3]
```diff
@@ -178,6 +178,17 @@ void run() throws Exception {
         }
     }
 
+    @Parameters(commandDescription = "Update the scale of transaction coordinators")
+    private class ScaleTransactionCoordinators extends CliCommand {
+        @Parameter(names = { "-r", "--replicas" }, description = "The scale of the transaction coordinators")
+        private int replicas;
```
Comment (Technoboy-) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java:184]: Which name is better, `partitions` or `replicas` ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 4]
```diff
@@ -432,4 +434,35 @@ protected void validateTopicName(String property, String namespace, String encod
...
         }
     }
+
+    protected void internalScaleTransactionCoordinators(int replicas, boolean authoritative) {
+        final int maxPartitions = pulsar().getConfig().getMaxNumPartitionsPerPartitionedTopic();
+        if (maxPartitions > 0 && replicas > maxPartitions) {
+            throw new RestException(Response.Status.NOT_ACCEPTABLE,
+                    "Number of partitions should be less than or equal to " + maxPartitions);
+        }
+        validateTopicOwnership(SystemTopicNames.TRANSACTION_COORDINATOR_ASSIGN, authoritative);
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:444]: validateTopicOwnership -> validateTopicOwnershipAsync

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 5]
```diff
@@ -432,4 +434,35 @@ protected void validateTopicName(String property, String namespace, String encod
...
+                        throw new RestException(Response.Status.NOT_ACCEPTABLE, "Number of partitions should "
+                                + "be more than the current number of transaction coordinator partitions");
+                    }
+                    return new PartitionedTopicMetadata(replicas);
+                }).thenAccept(r -> updatePartition.complete(null)).exceptionally(ex -> {
+                    updatePartition.completeExceptionally(ex.getCause());
+                    return null;
+                });
+        try {
+            updatePartition.get(DEFAULT_OPERATION_TIMEOUT_SEC, TimeUnit.SECONDS);
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:458]: New REST APIs should all use the async method.

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java (hunk 3 above)
Comment (liangyepianzhou) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java:184]: Maybe we don't need to tangle here, I have followed Penghui's suggestion to change from `partitions` to `replicas`.
Both names are fine. How to keep it like this?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java (hunk 4 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:444]: This is for authentication, just like `internalUpdatePartitionedTopic` in `PersistentTopicsBase`.
It should be sync.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java (hunk 5 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:458]: OK, I will abide by this specification

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 6]
```diff
@@ -432,4 +433,32 @@ protected void validateTopicName(String property, String namespace, String encod
...
+        if (maxPartitions > 0 && replicas > maxPartitions) {
+            throw new RestException(Response.Status.NOT_ACCEPTABLE,
+                    "Number of partitions should be less than or equal to " + maxPartitions);
+        }
+        validateSuperUserAccessAsync().thenAccept((ignore) -> {
+                    namespaceResources().getPartitionedTopicResources()
+                            .updatePartitionedTopicAsync(SystemTopicNames.TRANSACTION_COORDINATOR_ASSIGN, p -> {
+                                if (p.partitions >= replicas) {
+                                    throw new RestException(Response.Status.NOT_ACCEPTABLE,
+                                            "Number of partitions should "
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:448]:
```suggestion
                                            "Number of partitions should "
```
Number of transaction coordinators

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 7]
```diff
@@ -432,4 +433,32 @@ protected void validateTopicName(String property, String namespace, String encod
...
+                                            + "be more than the current number of transaction coordinator partitions");
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:449]:
```suggestion
                                            + "be more than the current number of transaction coordinator partitions");
```
be more than the currenst number of transaction coordinator

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/TransactionsImpl.java [hunk 8]
```diff
@@ -335,4 +338,17 @@ public TransactionPendingAckInternalStats getPendingAckInternalStats(String topi
...
     }
 
+    @Override
+    public void scaleTransactionCoordinators(int replicas) throws PulsarAdminException {
+         sync(() -> scaleTransactionCoordinatorsAsync(replicas));
+    }
+
+    @Override
+    public CompletableFuture<Void> scaleTransactionCoordinatorsAsync(int replicas) {
+        checkArgument(replicas > 0, "Number of partitions must be more than 0");
```
Comment (congbobo184) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/TransactionsImpl.java:348]: transaction coordinators

**Review by congbobo184 — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 9]
```diff
@@ -432,4 +433,27 @@ protected void validateTopicName(String property, String namespace, String encod
             throw new RestException(Response.Status.PRECONDITION_FAILED, "Topic name is not valid");
         }
     }
+
+    protected void internalScaleTransactionCoordinators(AsyncResponse asyncResponse, int replicas) {
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:437]: We should define this method as `protected CompletableFuture<Void> internalScaleTransactionCoordinators`,
which means we can't pass `AsyncResponse` into the method.

Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:437]: Yes, agree. It's a great practice.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 10]
```diff
@@ -432,4 +433,27 @@ protected void validateTopicName(String property, String namespace, String encod
...
+                            .exceptionally(ex -> {
+                                log.error("{} Failed to update the scale of transaction coordinators", clientAppId());
+                                resumeAsyncResponseExceptionally(asyncResponse, ex);
+                                return null;
+                            });
+                }).exceptionally(e -> {
+                    log.error("{} Failed to update the scale of transaction coordinators", clientAppId());
+                    resumeAsyncResponseExceptionally(asyncResponse, e);
+                    return null;
+                });
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:457]: Use `thenCompose().thenAccept().exceptionally()` to simplify the implementation.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java [hunk 11]
```diff
@@ -314,4 +316,25 @@ public void getPendingAckInternalStats(@Suspended final AsyncResponse asyncRespo
             resumeAsyncResponseExceptionally(asyncResponse, ex);
         }
     }
+
+    @POST
+    @Path("/transactionCoordinator/replicas")
+    @ApiResponses(value = {
+            @ApiResponse(code = 503, message = "This Broker is not configured "
+                    + "with transactionCoordinatorEnabled=true."),
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java:324]: Have you added this check in the implementation?
And should also add 

```
            @ApiResponse(code = 401, message = "This operation requires super-user access"),
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/v3/AdminApiTransactionTest.java [hunk 12]
```diff
@@ -519,6 +519,38 @@ public void testTransactionNotEnabled() throws Exception {
...
+    @Test
+    public void testUpdateTransactionCoordinatorNumber() throws Exception {
+        int coordinatorSize = 3;
+        pulsar.getPulsarResources()
+                .getNamespaceResources()
+                .getPartitionedTopicResources()
+                .createPartitionedTopic(SystemTopicNames.TRANSACTION_COORDINATOR_ASSIGN,
+                        new PartitionedTopicMetadata(coordinatorSize));
+        try {
+            admin.transactions().scaleTransactionCoordinators(coordinatorSize - 1);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/v3/AdminApiTransactionTest.java:538]: Could you please also add a test for auth enabled? only the super user can access this API.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 13]
```diff
@@ -432,4 +433,24 @@ protected void validateTopicName(String property, String namespace, String encod
...
+                                throw new RestException(Response.Status.NOT_ACCEPTABLE,
+                                        "Number of transaction coordinators should "
+                                                + "be more than the current number of transaction coordinator");
+                            }
+                            return new PartitionedTopicMetadata(replicas);
+                        }).thenAccept(r -> completableFuture.complete(null)))
+                .exceptionally(e -> {
+                    log.error("{} Failed to update the scale of transaction coordinators", clientAppId());
+                    completableFuture.completeExceptionally(e);
+                    return null;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:452]: Method `scaleTransactionCoordinators` already handled the exceptions, we don't need to handle them again here. Just return the `CompletableFuture`.

And for the log here, we don't need an error log, warn level log is enough.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 14]
```diff
@@ -432,4 +433,24 @@ protected void validateTopicName(String property, String namespace, String encod
             throw new RestException(Response.Status.PRECONDITION_FAILED, "Topic name is not valid");
         }
     }
+
+    protected CompletableFuture<Void> internalScaleTransactionCoordinators(int replicas) {
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:438]: We don't need to create the CompletableFuture here but complete it in another thread.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 15]
```diff
@@ -432,4 +433,24 @@ protected void validateTopicName(String property, String namespace, String encod
...
+        validateSuperUserAccessAsync()
+                .thenCompose((ignore)-> namespaceResources().getPartitionedTopicResources()
+                        .updatePartitionedTopicAsync(SystemTopicNames.TRANSACTION_COORDINATOR_ASSIGN, p -> {
+                            if (p.partitions >= replicas) {
+                                throw new RestException(Response.Status.NOT_ACCEPTABLE,
+                                        "Number of transaction coordinators should "
+                                                + "be more than the current number of transaction coordinator");
+                            }
+                            return new PartitionedTopicMetadata(replicas);
+                        }).thenAccept(r -> completableFuture.complete(null)))
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:448]: We can remove the `.thenAccept(r -> completableFuture.complete(null)))`, just return a CompletableFuture in `updatePartitionedTopicAsync`?

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java [hunk 16]
```diff
@@ -189,5 +200,6 @@ public CmdTransactions(Supplier<PulsarAdmin> admin) {
         jcommander.addCommand("transaction-in-pending-ack-stats", new GetTransactionInPendingAckStats());
         jcommander.addCommand("transaction-metadata", new GetTransactionMetadata());
         jcommander.addCommand("slow-transactions", new GetSlowTransactions());
+        jcommander.addCommand("scale-transactionCoordinators", new ScaleTransactionCoordinators());
```
Comment (codelipenghui) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java:203]: Need a test to cover here to make sure the CLI works.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java (hunk 11 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v3/Transactions.java:324]: `checkTransactionCoordinatorEnabled(); ` had been aaded, and I added auth and auth test this time.

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 17]
```diff
@@ -432,4 +433,17 @@ protected void validateTopicName(String property, String namespace, String encod
             throw new RestException(Response.Status.PRECONDITION_FAILED, "Topic name is not valid");
         }
     }
+
+    protected CompletableFuture<Void> internalScaleTransactionCoordinators(int replicas) {
+        return validateSuperUserAccessAsync()
+                .thenCompose((ignore)-> namespaceResources().getPartitionedTopicResources()
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:439]: (ignore) ->  __

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**